src/modules/pulumi.rs

```diff
@@ -0,0 +1,295 @@
+            Some(PathBuf::from("/home/sweet/home/.pulumi"))
+        );
+        context.env.insert("PULUMI_HOME", "/a/dir".to_string());
+        assert_eq!(pulumi_home_dir(&context), Some(PathBuf::from("/a/dir")))
+    }
+
+    #[test]
+    fn test_get_pulumi_workspace() {
+        let mut context = Context::new(ArgMatches::default());
+        context.env.insert("HOME", "/home/sweet/home".to_string());
+        let name = "foobar";
+        let project_file = PathBuf::from("/hello/Pulumi.yaml");
+        assert_eq!(
+            get_pulumi_workspace(&context, name, &project_file),
+            Some("/home/sweet/home/.pulumi/workspaces/foobar-test-workspace.json")
+                .map(PathBuf::from)
+        );
+    }
+
+    #[test]
+    fn version_render() -> io::Result<()> {
+        let dir = tempfile::tempdir()?;
+        let pulumi_file = File::create(dir.path().join("Pulumi.yaml"))?;
+        pulumi_file.sync_all()?;
+        let rendered = ModuleRenderer::new("pulumi")
+            .path(dir.path())
+            .config(toml::toml! {
+                [pulumi]
+                format = "with [$version]($style) "
+            })
+            .collect();
+        dir.close()?;
+        let expected = format!("with {} ", Color::Fixed(5).bold().paint("v1.2.3-ver"));
+
+        assert_eq!(expected, rendered.expect("a result"));
+        Ok(())
+    }
+
+    #[test]
+    /// This test confirms a full render. This means finding a Pulumi.yml file,
+    /// tracing back to the backing workspace settings file, and printing the
+    /// stack name.
+    fn render_valid_paths() -> io::Result<()> {
+        use io::Write;
+        let dir = tempfile::tempdir()?;
+        let root = std::fs::canonicalize(dir.path())?;
+        let mut yaml = File::create(root.join("Pulumi.yml"))?;
+        yaml.write_all("name: starship\nruntime: nodejs\ndescription: A thing\n".as_bytes())?;
+        yaml.sync_all()?;
+
+        let workspace_path = root.join(".pulumi").join("workspaces");
+        let _ = std::fs::create_dir_all(&workspace_path)?;
+        let workspace_path = &workspace_path.join("starship-test-workspace.json");
+        let mut workspace = File::create(&workspace_path)?;
+        serde_json::to_writer_pretty(
+            &mut workspace,
+            &serde_json::json!(
+                {
+                    "stack": "launch"
+                }
+            ),
+        )?;
+        workspace.sync_all()?;
+        let rendered = ModuleRenderer::new("pulumi")
+            .path(root.clone())
+            .logical_path(root.clone())
+            .config(toml::toml! {
+                [pulumi]
+                format = "in [$symbol($stack)]($style) "
+            })
+            .env("HOME", root.to_str().unwrap())
+            .collect();
+        let expected = format!("in {} ", Color::Fixed(5).bold().paint("🚀 launch"));
+        assert_eq!(expected, rendered.expect("a result"));
+        dir.close()?;
+        Ok(())
+    }
```

Comment:
I noticed you pushed an update that allowed you to deal with empty Pulumi configs. Is there any way to write a test for that? (E.g. by creating an empty file masquerading as a config and then running that). Or are there other problems that I haven't forseen?

Reply:
We can (and I will) write such a test. It will be included with the next commit.